packages/lodestar/src/chain/genesis/genesis.ts

```diff
@@ -28,16 +28,167 @@ import {
+    this.config = config;
+    this.eth1 = eth1;
+    this.db = db;
+    this.logger = logger;
+    this.deposits = [];
+  }
+
+
+  /**
+   * This processes an eth1 block and check if it can form genesis state of eth1.
+   */
+  public isFormingValidGenesisState = async (events: IDepositEvent[]): Promise<boolean> => {
+    const block = await this.eth1.getBlock(events[0].blockNumber);
+    applyTimestamp(this.config, this.state, block.timestamp);
+    applyEth1BlockHash(this.config, this.state, fromHexString(block.hash));
+    const isValid = isValidGenesisState(this.config, this.state);
+    this.logger.verbose(`genesis: Process block ${block.number}, valid genesis state=${isValid}`);
+    if (isValid) {
+      const eth1Data = {
+        blockHash: fromHexString(block.hash),
+        depositRoot: this.depositTree.tree().root,
+        depositCount: events[events.length - 1].index + 1,
+      };
+
+      await Promise.all([
+        // proposing data
+        this.db.depositData.batchPut(events.map((depositEvent) => ({
+          key: depositEvent.index,
+          value: depositEvent,
+        }))),
+        this.db.eth1Data.put(block.timestamp, eth1Data)
+      ]);
+    }
+    return isValid;
+  };
+
+  public processDepositEvents(): ((source: AsyncIterable<IDepositEvent[]>) => AsyncIterable<IDepositEvent[]>) {
+    const {logger, doProcessDepositEvents, config} = this;
+    return ((source: AsyncIterable<IDepositEvent[]>): AsyncIterable<IDepositEvent[]> => {
+      return async function* () {
+        for await (const depositEvents of source) {
+          const lastBlockNumber = depositEvents[depositEvents.length - 1].blockNumber;
+          const firstBlockNumber = depositEvents[0].blockNumber;
+          for (let blockNumber = firstBlockNumber; blockNumber <= lastBlockNumber; blockNumber++) {
+            const blockDepositEvents = depositEvents.filter(event => event.blockNumber === blockNumber);
+            if (blockDepositEvents.length > 0) {
+              await doProcessDepositEvents(blockDepositEvents);
+              const depositCount = blockDepositEvents[blockDepositEvents.length - 1].index + 1;
+              logger.info(`genesis: Found ${blockDepositEvents.length} deposit events for block ${blockNumber}, ` +
+                `depositCount=${depositCount}`);
+              if (depositCount >= config.params.MIN_GENESIS_ACTIVE_VALIDATOR_COUNT) {
+                yield blockDepositEvents;
+              }
+            }
+          }
+        }
+      }();
+    });
+  }
+
+  public genesis = async (): Promise<TreeBacked<BeaconState>> => {
+    await this.initialize();
+    const eth1DataStream = await this.eth1.getDepositEventsFromBlock(true, undefined);
+    const {isFormingValidGenesisState: processSingleBlock, state} = this;
+    const foundGenesis = this.eth1.foundGenesis.bind(this.eth1);
+    return await pipe(eth1DataStream,
+      this.processDepositEvents(),
```

Comment:
It makes more sense for this to have eth1Blocks as source than deposit events

Reply:
we want to process deposit events here by applying deposits to the cached state. Once we have enough deposits, we'll group deposit events by block, get eth1block and do `processSingleBlock`

Reply:
why not have one pipe where you process just some batched eth1 deposit events source and when you have enough, end pipe and start another where source are eth1 blocks which you then process?

Reply:
we want both block data (mainly hash + timestamp) and deposit events in order to save proposing data (eth1Data + deposit data). I'm thinking about moving the persistence of proposing away from here to make it single responsibility.